Release notes for 3.0 (#6888)
dotnet/machinelearning: docs/release-notes/3.0/release-3.0.0.md

# [ML.NET](http://dot.net/ml) 3.0.0

## **New Features**
- **Add the ability to use Object Detection using TorchSharp** ([#6605](https://github.com/dotnet/machinelearning/pull/6605)) - We have added a new deep learning model back by TorchSharp that lets you fine tune your own Object Detection model!
- **Add SamplingKeyColumnName to AutoMLExperiment API** ([#6649](https://github.com/dotnet/machinelearning/pull/6649)) - You can now set the SamplingKeyColumnName when you are using AutoML. Thanks @torronen!
- **Add Object Detection to AutoML Sweeper** ([#6633](https://github.com/dotnet/machinelearning/pull/6633)) - Added Object Detection to the AutoML Sweeper so now they can be used together.
- **Add String Vector support to DataFrame** ([#6628](https://github.com/dotnet/machinelearning/pull/6628)) - Adds support for String Vectors in DataFrame. This also allows for Better IDataView <-> DataFrame conversions.

> - **Add AutoZero tuner to BinaryClassification** ([#6615](https://github.com/dotnet/machinelearning/pull/6615)) - Can now user AutoZero tuner in AutoML Binary Classification experiments.

Reviewer: >user [](http://example.com/codeflow?start=122&length=4)

use?

Reviewer 2: ```suggestion
- **Add AutoZero tuner to BinaryClassification** ([#6615](https://github.com/dotnet/machinelearning/pull/6615)) - You can now use AutoZero tuner in AutoML Binary Classification experiments.
```

Author: Good catch.

- **Add in ability to have pre-defined weights for ngrams** ([#6458](https://github.com/dotnet/machinelearning/pull/6458)) - If you know the weights of your NGrams already you can now directly provide that.
- **Add SentenceSimilarity sweepable estimator in AutoML** ([#6445](https://github.com/dotnet/machinelearning/pull/6445)) - Can now use SentenceSimilarity with the sweepable estimator.
- **Add VBufferDataFrameCoumn to DataFrame** ([#6409](https://github.com/dotnet/machinelearning/pull/6409)) - Now DataFrame can support the VBuffer from ML.NET so the IDataView <-> DataFrame conversion can work with those types.
- **Added ADO.NET importing/exporting functionality to DataFrame** ([#5975](https://github.com/dotnet/machinelearning/pull/5975)) - Can now use ADO.NET import/export with DataFrames. Thanks @andrei-faber!

> - **Added in native binaries for Windows Arm64** ([#6813](https://github.com/dotnet/machinelearning/pull/6813))

Reviewer 2: ```suggestion
- **Added native binaries for Windows Arm64** ([#6813](https://github.com/dotnet/machinelearning/pull/6813))
```

We should add a note for what this offers. I presume it's improved performance on Arm64; is that correct?

Author: Added.

- **Switches some computational code to use the new Tensor Primitives package** ([#6875](https://github.com/dotnet/machinelearning/pull/6875))

> - **Add QA sweepable estimator in AutoML ** ([#6781](https://github.com/dotnet/machinelearning/pull/6781))

Reviewer 3: This isn't rendering right in markdown. Remove the space after AutoML to get the bold working properly.

Reviewer 2: ```suggestion
- **Add QA sweepable estimator in AutoML** ([#6781](https://github.com/dotnet/machinelearning/pull/6781))
```

Author: Good catch.

- **Add default search space for standard trainers** ([#6576](https://github.com/dotnet/machinelearning/pull/6576)) - Added a default search space for all standard trainers so users have reasonable default values.
- **Adding more metrics to BinaryClassification Experiment** ([#6571](https://github.com/dotnet/machinelearning/pull/6571))
- **Add checkAlive in NasBertTrainer** ([#6546](https://github.com/dotnet/machinelearning/pull/6546)) - Now we check between batches if cancellation was requested and stop processing if so.
- **OneDAL - Fallback to default implementation** ([#6538](https://github.com/dotnet/machinelearning/pull/6538)) - If you specify you want to use OneDal but something happens that prevents you from using it, like it can't find the binaries/etc, it will auto default back to the normal implementation instead of crashing.
- **Add addKeyValueAnnotationsAsText flag in AutoML** ([#6535](https://github.com/dotnet/machinelearning/pull/6535))
- **Add continuous resource monitoring to AutoML.IMonitor** ([#6520](https://github.com/dotnet/machinelearning/pull/6520)) - Thanks @andrasfuchs!

> - **Update WebClient to HttpClient implementations** ([#6476](https://github.com/dotnet/machinelearning/pull/6476)) - Update a usage of WebClient to HttpClient since WebClient is now deprecated. Thanks @rgesteve!

Reviewer: >a [](http://example.com/codeflow?start=125&length=1)

the

Reviewer 2: ```suggestion
- **Update WebClient to HttpClient implementations** ([#6476](https://github.com/dotnet/machinelearning/pull/6476)) - Update usage of WebClient to HttpClient since WebClient is now deprecated. Thanks @rgesteve!
```

- **Set AutoML trial to unsuccess if trial loss is nan/inf** ([#6430](https://github.com/dotnet/machinelearning/pull/6430)) - Now trial will be marked as unsuccesssful if the loss is an invalid number.
- **Add diskConvert option in fast tree search space** ([#6316](https://github.com/dotnet/machinelearning/pull/6316))
- **Avoid Boxing/Unboxing on accessing elements of VBufferDataFrameColumn** ([#6867](https://github.com/dotnet/machinelearning/pull/6867)) and ([#6865](https://github.com/dotnet/machinelearning/pull/6865)) - Thanks @asmirnov82!

> - **Update LightGBM to version 3.X.X from 2.X.X ([#6880](https://github.com/dotnet/machinelearning/pull/6880))

Reviewer 3: Needs a closing ** after 2.X.X

Author: Fixed.

> - **Update LightGBM to version 3.X.X from 2.X.X ([#6880](https://github.com/dotnet/machinelearning/pull/6880))

Reviewer 2: ```suggestion
- **Update LightGBM to version 3.X.X from 2.X.X** ([#6880](https://github.com/dotnet/machinelearning/pull/6880))
```

Author: Fixed.

- **Implement vectorized binary arithmetic operations for DataFrames** ([#6854](https://github.com/dotnet/machinelearning/pull/6854)) - Thanks @asmirnov82!

> - **Upgrade .NET Interactive** ([#6857](https://github.com/dotnet/machinelearning/pull/6857)) - Thanks @colombod

Reviewer 2: ```suggestion
- **Upgrade .NET Interactive** ([#6857](https://github.com/dotnet/machinelearning/pull/6857)) - Thanks @colombod!
```

- **Improve performance of column cloning inside DataFrame arithmetics** ([#6814](https://github.com/dotnet/machinelearning/pull/6814)) - Thanks @asmirnov82!
- **Add performance benchmarks for dataframe arithmetic operations** ([#6827](https://github.com/dotnet/machinelearning/pull/6827)) - Thanks @asmirnov82!

> - **Simplify tt files for PrimitiveDataFrameColumnAritmetics** ([#6830](https://github.com/dotnet/machinelearning/pull/6830)) - Thanks @asmirnov82!
> - **Update dependencies** ([#6837](https://github.com/dotnet/machinelearning/pull/6837))
> - **update interactive kernel version** ([#6836](https://github.com/dotnet/machinelearning/pull/6836))
> - **Improve performance of DataFrame binary comparison operations** ([#6869](https://github.com/dotnet/machinelearning/pull/6869)) - Thanks @asmirnov82!

Reviewer 2: ```suggestion
- **Simplify tt files for PrimitiveDataFrameColumnAritmetics** ([#6830](https://github.com/dotnet/machinelearning/pull/6830)) - Thanks @asmirnov82!
- **Improve performance of DataFrame binary comparison operations** ([#6869](https://github.com/dotnet/machinelearning/pull/6869)) - Thanks @asmirnov82!
```

I moved the dependency updates down to the bottom with others like them.

> - **Allow to define CultureInfo for parsing values on reading DataFrame from csv** ([#6782](https://github.com/dotnet/machinelearning/pull/6782)) - Thanks @asmirnov82!

Reviewer 2: ```suggestion
- **Allow a CultureInfo to be used for parsing CSV values into DataFrame** ([#6782](https://github.com/dotnet/machinelearning/pull/6782)) - Thanks @asmirnov82!
```

## **Bug Fixes**
- **Fix DataFrame ToString** ([#6673](https://github.com/dotnet/machinelearning/pull/6673)) - Use correct alignment for columns to produce readable output when columns have longer names. Thanks @asmirnov82!
- **Fix DataFrame null math** ([#6661](https://github.com/dotnet/machinelearning/pull/6661)) - Fixes max in DataFrame columns when there are null values to match what Pandas does.
- **Clean up PrimitiveColumnContainer** ([#6656](https://github.com/dotnet/machinelearning/pull/6656)) - Cleaned up the code in PrimitiveColumnContainer so its more correct and easier to use.
- **Fix Apply in PrimitiveColumnContainer** ([#6642](https://github.com/dotnet/machinelearning/pull/6642)) - Fixes the Apply method so it no longer changes the source column. Thanks @janholo!
- **Fix datetime null error** ([#6627](https://github.com/dotnet/machinelearning/pull/6627)) - Fixes loading a null datetime from a database so it now returns correctly instead of throwing an error.
- **Fix AggregateTrainingStopManager is trying to cancel disposed tokens** ([#6612](https://github.com/dotnet/machinelearning/pull/6612)) - Will no longer try and cancel already disposed tokens.

> - **Fix tostring bug for sweepable pipeline** ([#6610](https://github.com/dotnet/machinelearning/pull/6610))

Reviewer 2: ```suggestion
- **Fix ToString bug for sweepable pipeline** ([#6610](https://github.com/dotnet/machinelearning/pull/6610))
```

- **Change Test to Validate in Dataset manager** ([#6599](https://github.com/dotnet/machinelearning/pull/6599))
- **Fixed System.OperationCanceledException when calling experimentResult.BestRun.Estimator.Fit** ([#6572](https://github.com/dotnet/machinelearning/pull/6572))
- **Fixed cancellation bug in SweepablePipelineRunner && Fixed object null exception in AutoML v1.0 regression API** ([#6560](https://github.com/dotnet/machinelearning/pull/6560))

> - **Fixed one dal dispatching issues** ([#6547](https://github.com/dotnet/machinelearning/pull/6547)) - OneDal now dispatches correctly.

Reviewer 2: ```suggestion
- **Fixed OneDal dispatching issues** ([#6547](https://github.com/dotnet/machinelearning/pull/6547)) - OneDal now dispatches correctly.
```

- **PrimitiveDataFrameColumn.Clone method crashes when is used with IEnumerable mapIndices argument** ([#6822](https://github.com/dotnet/machinelearning/pull/6822)) - Thanks @asmirnov82!
- **Addresses #6533, OneDal Index was outside the bounds of the array.** ([#6838](https://github.com/dotnet/machinelearning/pull/6838)) - Thanks @rgesteve!
- **Fix wrong type conversion on PrimitiveDataFrameColumn** ([#6834](https://github.com/dotnet/machinelearning/pull/6834)) - Thanks @novelhawk!
- **Append dataframe rows based on column names** ([#6808](https://github.com/dotnet/machinelearning/pull/6808)) - Thanks @asmirnov82!
- **Fix inconsistent null handling in DataFrame Arithmetics** ([#6770](https://github.com/dotnet/machinelearning/pull/6770)) - Thanks @asmirnov82!
- **Fix DataFrame.LoadCsv can not load CSV with duplicate column names** ([#6772](https://github.com/dotnet/machinelearning/pull/6772)) - Thanks @asmirnov82!
- **Fix issue with addIndexColumn in DataFrame.LoadCsv** ([#6769](https://github.com/dotnet/machinelearning/pull/6769)) - Thanks @asmirnov82!

> - **fix issue #2718 from model builder** ([#6768](https://github.com/dotnet/machinelearning/pull/6768))

Reviewer 2: ```suggestion
- **Fix text classification InvocationException during cross-validation, issue #2718 from model builder** ([#6768](https://github.com/dotnet/machinelearning/pull/6768))
```

- **Fix DataFrame bounds checking on indexing elements** ([#6681](https://github.com/dotnet/machinelearning/pull/6681)) - Thanks @asmirnov82!
- **Reset DataFrame.RowCount to zero, when DataFrame is empty** ([#6698](https://github.com/dotnet/machinelearning/pull/6698)) - Thanks @asmirnov82!
- **Stop shuffle rows in ITrainValidationDatasetManager** ([#6742](https://github.com/dotnet/machinelearning/pull/6742))
- **SMAC - ignore fail trial during initialize** ([#6738](https://github.com/dotnet/machinelearning/pull/6738))
- **Fix non-thread-safe use of Random in tokenizers** ([#6695](https://github.com/dotnet/machinelearning/pull/6695))
- **Fixing license** ([#6689](https://github.com/dotnet/machinelearning/pull/6689)) and  ([#6690](https://github.com/dotnet/machinelearning/pull/6690))

## **Build / Test updates**
- **Remove MSIL Check for TorchSharp** ([#6658](https://github.com/dotnet/machinelearning/pull/6658)) - Removes the MSIL check for TorchSharp while we figure out how we want to correctly handle this.
- **Change code coverage build pool** ([#6647](https://github.com/dotnet/machinelearning/pull/6647)) - Changed codecoverage build pool so the builds are faster and more stable.

> - **Update AutoMLExperimentTests.cs to fix timeout error** ([#6638](https://github.com/dotnet/machinelearning/pull/6638))
> - **Update FabricBot config** ([#6619](https://github.com/dotnet/machinelearning/pull/6619))
> - **Libraries area pod updates March 2023** ([#6607](https://github.com/dotnet/machinelearning/pull/6607))

Reviewer 2: ```suggestion
- **Update AutoMLExperimentTests.cs to fix timeout error** ([#6638](https://github.com/dotnet/machinelearning/pull/6638))
```

> - **Update dependencies from dotnet/arcade** ([#6566 & #6518 & #6451 & #6439](https://github.com/dotnet/machinelearning/pull/6566))

Reviewer 2: ```suggestion
- **update interactive kernel version** ([#6836](https://github.com/dotnet/machinelearning/pull/6836))
- **Update dependencies** ([#6837](https://github.com/dotnet/machinelearning/pull/6837))
- **Update dependencies from dotnet/arcade** ([#6566 & #6518 & #6451 & #6439](https://github.com/dotnet/machinelearning/pull/6566))
```

- **Mac python fix** ([#6549](https://github.com/dotnet/machinelearning/pull/6549))
- **Moving onedal nuget download from onedal to native where its needed for building** ([#6527](https://github.com/dotnet/machinelearning/pull/6527))

> - **New os image for official builds** ([#6467](https://github.com/dotnet/machinelearning/pull/6467))
> - **FabricBot: Do not add needs-further-triage to issues still marked as untriaged** ([#6722](https://github.com/dotnet/machinelearning/pull/6722))
> - **FabricBot: Remove area pod project board automation** ([#6881](https://github.com/dotnet/machinelearning/pull/6881))
> - **Removed codecov token** ([#6811](https://github.com/dotnet/machinelearning/pull/6811))
> - **Removed deprecated yosemite brew** ([#6805](https://github.com/dotnet/machinelearning/pull/6805))

Reviewer 2: ```suggestion
- **New os image for official builds** ([#6467](https://github.com/dotnet/machinelearning/pull/6467))
- **Removed deprecated yosemite brew** ([#6805](https://github.com/dotnet/machinelearning/pull/6805))
```
